src/composed/message/parser.rs

```diff
@@ -106,21 +121,25 @@ fn next<I: Iterator<Item = Packet>>(packets: &mut Peekable<I>) -> Option<Result<
                 return match packet.try_into() {
                     Ok(p) => {
                         let one_pass_signature = Some(p);
-                        let m = next(packets.by_ref());
-                        let message = if let Some(Err(err)) = m {
-                            return Some(Err(err));
-                        } else {
-                            m.map(|m| Box::new(m.expect("already checked")))
+
+                        let message = match next(packets.by_ref()) {
+                            Some(Ok(m)) => Some(Box::new(m)),
+                            Some(Err(err)) => return Some(Err(err)),
+                            None => None,
                         };
 
-                        let signature =
-                            if packets.peek().map(|p| p.tag() == Tag::Signature) == Some(true) {
-                                packets.next().expect("peeked").try_into().expect("peeked")
-                            } else {
-                                return Some(Err(format_err!(
-                                    "missing signature for, one pass signature"
-                                )));
-                            };
+                        let signature = if packets.peek().map_or(false, |res| {
```

Comment:
I think all these `peek` calls could be simplified with this: https://doc.rust-lang.org/stable/std/iter/struct.Peekable.html#method.next_if not sure why I didn't use this back then.